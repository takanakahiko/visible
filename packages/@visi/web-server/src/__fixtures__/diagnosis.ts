import { Diagnosis, Status } from '../domain/models';
import { source } from './source';

export const diagnosis = Diagnosis.from({
  id: '08eecb12-75a1-4798-aca2-f9e919b1fd56',
  status: Status.STARTED,
  url: 'https://exmaple.com',
  screenshot: 'https://example.com',
  sources: [source],
  doneCount: 1,
  totalCount: 1,
  createdAt: new Date('2020-05-17T08:28:59.186Z'),
  updatedAt: new Date('2020-05-17T08:28:59.186Z'),
  waitingCountAtCreation: 0,
  completeCountAtCreation: 0,
});

export const minimalDiagnosis = Diagnosis.from({
  id: '08eecb12-75a1-4798-aca2-f9e919b1fd56',
  status: Status.STARTED,
  url: 'https://exmaple.com',
  screenshot: 'https://example.com',
  sources: [],
  doneCount: 0,
  totalCount: 0,
  createdAt: new Date('2020-05-17T08:28:59.186Z'),
  updatedAt: new Date('2020-05-17T08:28:59.186Z'),
  waitingCountAtCreation: 0,
  completeCountAtCreation: 0,
});
