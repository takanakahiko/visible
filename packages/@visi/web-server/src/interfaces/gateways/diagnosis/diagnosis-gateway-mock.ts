import { injectable, unmanaged } from 'inversify';
import { Observable, Subject } from 'rxjs';

import { DiagnosisRepository } from '../../../application/repositories';
import { Diagnosis } from '../../../domain/models';

@injectable()
export class DiagnosisGatewayMock implements DiagnosisRepository {
  private readonly update$ = new Subject<Diagnosis>();

  constructor(
    @unmanaged()
    private readonly diagnoses = new Map<string, Diagnosis>(),
  ) {}

  async find(ids: string[]): Promise<Diagnosis[]> {
    return ids
      .map((id) => this.diagnoses.get(id))
      .filter((value): value is Diagnosis => value != null);
  }

  async save(diagnosis: Diagnosis): Promise<Diagnosis> {
    this.diagnoses.set(diagnosis.id, diagnosis);
    const result = this.diagnoses.get(diagnosis.id);
    if (!result) throw new Error('No entry found');
    return result;
  }

  async delete(id: string): Promise<string> {
    this.diagnoses.delete(id);
    return id;
  }

  async queue(): Promise<void> {
    return;
  }

  async publish(diagnosis: Diagnosis): Promise<void> {
    this.update$.next(diagnosis);
  }

  subscribe(): Observable<Diagnosis> {
    return this.update$;
  }
}
