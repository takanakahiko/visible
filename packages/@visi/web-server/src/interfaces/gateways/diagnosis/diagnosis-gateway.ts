import { validateOrReject } from 'class-validator';
import { inject, injectable } from 'inversify';
import { Subject } from 'rxjs';
import { filter } from 'rxjs/operators';
import { Connection } from 'typeorm';

import { DiagnosisRepository } from '../../../application/repositories';
import { Diagnosis } from '../../../domain/models';
import { Logger } from '../../../domain/services';
// eslint-disable-next-line
import { ProcessDiagnosisJob, PublishDiagnosisJob } from '../../../frameworks/jobs';
import { TYPES } from '../../../types';
import { DiagnosisTable } from './diagnosis-table';

@injectable()
export class DiagnosisGateway implements DiagnosisRepository {
  private readonly publish$: Subject<Diagnosis>;

  constructor(
    @inject(TYPES.Logger)
    private readonly logger: Logger,

    @inject(TYPES.Connection)
    private readonly connection: Connection,

    @inject(TYPES.PublishDiagnosisJob)
    private readonly publishDiagnosis: PublishDiagnosisJob,

    @inject(TYPES.ProcessDiagnosisJob)
    private readonly processDiagnosis: ProcessDiagnosisJob,
  ) {
    this.publish$ = new Subject<Diagnosis>();
    this.publishDiagnosis.queue.process((job, done) => {
      const diagnosis = Diagnosis.from(JSON.parse(job.data));
      this.publish$.next(diagnosis);
      done();
    });
  }

  async find(ids: string[]) {
    const diagnoses = await this.connection
      .getRepository(DiagnosisTable)
      .findByIds(ids, {
        relations: [
          'reports',
          'reports.rule',
          'reports.pointers',
          'reports.pointers.source',
        ],
      });

    this.logger.debug(diagnoses);
    this.logger.debug(diagnoses.length);
    if (!diagnoses.length) throw new Error('Entry not found');

    return diagnoses.map((diagnosis) => diagnosis.toDomain());
  }

  private async findOne(id: string) {
    return this.connection
      .getRepository(DiagnosisTable)
      .findOne(id, {
        relations: [
          'reports',
          'reports.rule',
          'reports.pointers',
          'reports.pointers.source',
        ],
      })
      .then((result) => result?.toDomain());
  }

  async save(diagnosis: Diagnosis) {
    await validateOrReject(diagnosis);
    await this.connection
      .getRepository(DiagnosisTable)
      .save(DiagnosisTable.fromDomain(diagnosis));
    const result = await this.findOne(diagnosis.id);
    if (result == null) throw new Error('Save failed');
    return result;
  }

  async delete(id: string) {
    await this.connection.getRepository(DiagnosisTable).delete(id);
    return id;
  }

  async queue(diagnosis: Diagnosis) {
    await this.processDiagnosis.queue.add(diagnosis.toJSON());
  }

  subscribe(id: string) {
    this.logger.debug(`[Repository] Some client has subscribed to ${id}`);
    return this.publish$.pipe(filter((diagnosis) => diagnosis.id === id));
  }

  async publish(diagnosis: Diagnosis) {
    this.logger.debug(`[Repository] Publishing ${diagnosis.id}`);
    await this.publishDiagnosis.queue.add(diagnosis.toJSON());
  }
}
