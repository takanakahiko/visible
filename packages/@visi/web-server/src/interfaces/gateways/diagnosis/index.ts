export * from './diagnosis-db-entity';
export * from './diagnosis-gateway-mock';
export * from './diagnosis-gateway';
