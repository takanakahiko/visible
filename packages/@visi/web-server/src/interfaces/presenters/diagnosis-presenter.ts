import { inject, injectable } from 'inversify';

import { Diagnosis, Status } from '../../domain/models';
import { SourcePresenter } from './source-presenter';
import { API } from './types';

@injectable()
export class DiagnosisPresenter {
  constructor(
    @inject(SourcePresenter)
    private readonly sourcePresenter: SourcePresenter,
  ) {}

  transformStatus(status: Status): API.Status {
    switch (status) {
      case Status.QUEUED:
        return API.Status.Queued;
      case Status.DONE:
        return API.Status.Done;
      case Status.FAILED:
        return API.Status.Failed;
      case Status.PROCESSING:
        return API.Status.Processing;
      case Status.STARTED:
        return API.Status.Started;
    }
  }

  run(diagnosis: Diagnosis): API.Diagnosis {
    return {
      id: diagnosis.id,
      url: diagnosis.url,
      status: this.transformStatus(diagnosis.status),
      screenshot: diagnosis.screenshot,
      sources: diagnosis.sources.map((source) =>
        this.sourcePresenter.run(source),
      ),
      doneCount: diagnosis.doneCount,
      totalCount: diagnosis.totalCount,
      createdAt: diagnosis.createdAt,
      updatedAt: diagnosis.updatedAt,
    };
  }
}
