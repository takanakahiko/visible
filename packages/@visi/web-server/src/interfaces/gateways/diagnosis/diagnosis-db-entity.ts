import { Column, Entity, Index, OneToMany, PrimaryColumn } from 'typeorm';

import { Diagnosis, Status } from '../../../domain/models';
import { SourceDBEntity } from './source-db-entity';

@Entity('diagnosis')
export class DiagnosisDBEntity {
  @PrimaryColumn('uuid')
  id!: string;

  @Column('varchar', { length: 255 })
  status!: Status;

  @Column('text', { nullable: true })
  screenshot?: string;

  @Index()
  @Column('text')
  url!: string;

  @Column('int')
  doneCount!: number;

  @Column('int')
  totalCount!: number;

  @Index()
  @Column('timestamp')
  createdAt!: Date;

  @Index()
  @Column('timestamp')
  updatedAt!: Date;

  @Column('int', { default: 0 })
  waitingCountAtCreation!: number;

  @Column('int', { default: 0 })
  completeCountAtCreation!: number;

  @OneToMany(() => SourceDBEntity, (source) => source.diagnosis)
  readonly sources?: SourceDBEntity[];

  static fromDomain(diagnosis: Diagnosis): DiagnosisDBEntity {
    const entity = new DiagnosisDBEntity();
    entity.id = diagnosis.id;
    entity.url = diagnosis.url;
    entity.status = diagnosis.status;
    entity.screenshot = diagnosis.screenshot;
    entity.doneCount = diagnosis.doneCount;
    entity.totalCount = diagnosis.totalCount;
    entity.createdAt = diagnosis.createdAt;
    entity.updatedAt = diagnosis.updatedAt;
    entity.waitingCountAtCreation = diagnosis.waitingCountAtCreation;
    entity.completeCountAtCreation = diagnosis.completeCountAtCreation;
    return entity;
  }

  toDomain(): Diagnosis {
    if (this.sources == null) {
      throw new TypeError(
        'No value specified for sources property. You may have forgot JOIN?',
      );
    }

    return Diagnosis.from({
      id: this.id,
      status: this.status,
      screenshot: this.screenshot,
      url: this.url,
      doneCount: this.doneCount,
      totalCount: this.totalCount,
      sources: this.sources.map((sources) => sources.toDomain()),
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      waitingCountAtCreation: this.waitingCountAtCreation,
      completeCountAtCreation: this.completeCountAtCreation,
    });
  }
}
