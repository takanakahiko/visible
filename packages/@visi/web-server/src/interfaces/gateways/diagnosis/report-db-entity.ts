import { Column, Entity, ManyToOne, PrimaryColumn } from 'typeorm';

import { Location, Outcome, Report } from '../../../domain/models';
import { RuleDBEntity } from '../rule';
import { SourceDBEntity } from './source-db-entity';

@Entity('report')
export class ReportDBEntity {
  @PrimaryColumn('uuid')
  id!: string;

  @Column('varchar', { length: 255 })
  outcome!: Outcome;

  @Column('varchar', { length: 255, nullable: true })
  target?: string;

  @Column('varchar', { length: 255, nullable: true })
  message?: string;

  @Column('uuid')
  sourceId!: string;

  @Column('uuid')
  ruleId!: string;

  @Column('json', { nullable: true })
  location?: Location;

  @Column('text', { nullable: true })
  diffHunk?: string;

  @Column('varchar', { length: 255, nullable: true })
  screenshot?: string;

  @ManyToOne(() => SourceDBEntity, { onDelete: 'CASCADE' })
  readonly source?: SourceDBEntity;

  @ManyToOne(() => RuleDBEntity)
  readonly rule?: RuleDBEntity;

  static fromDomain(report: Report): ReportDBEntity {
    const entity = new ReportDBEntity();
    entity.id = report.id;
    entity.outcome = report.outcome;
    entity.target = report.target;
    entity.message = report.message;
    entity.sourceId = report.sourceId;
    entity.ruleId = report.ruleId;
    entity.diffHunk = report.diffHunk;
    entity.screenshot = report.screenshot;
    entity.location = report.location;
    return entity;
  }

  toDomain(): Report {
    return Report.from({
      id: this.id,
      outcome: this.outcome,
      ruleId: this.ruleId,
      target: this.target,
      message: this.message,
      sourceId: this.sourceId,
      diffHunk: this.diffHunk,
      screenshot: this.screenshot,
      location: this.location,
    });
  }
}
