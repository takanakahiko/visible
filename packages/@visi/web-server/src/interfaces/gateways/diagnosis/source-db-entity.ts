import { Column, Entity, ManyToOne, OneToMany, PrimaryColumn } from 'typeorm';

import { Source } from '../../../domain/models';
import { DiagnosisDBEntity } from './diagnosis-db-entity';
import { ReportDBEntity } from './report-db-entity';

@Entity('source')
export class SourceDBEntity {
  @PrimaryColumn('uuid')
  id!: string;

  @Column('text')
  content!: string;

  @Column('varchar', { length: 255, nullable: true })
  url?: string;

  @Column('uuid')
  diagnosisId!: string;

  @ManyToOne(() => DiagnosisDBEntity, (diagnosis) => diagnosis.sources)
  readonly diagnosis?: DiagnosisDBEntity;

  @OneToMany(() => ReportDBEntity, (report) => report.source)
  readonly reports?: ReportDBEntity[];

  static fromDomain(source: Source): SourceDBEntity {
    const entity = new SourceDBEntity();
    entity.id = source.id;
    entity.content = source.content;
    entity.url = source.url;
    entity.diagnosisId = source.diagnosisId;
    return entity;
  }

  toDomain(): Source {
    if (this.reports == null) {
      throw new TypeError(
        'No value specified for reports property. You may have forgot JOIN?',
      );
    }

    return Source.from({
      id: this.id,
      content: this.content,
      url: this.url,
      diagnosisId: this.diagnosisId,
      reports: this.reports.map((report) => report.toDomain()),
    });
  }
}
