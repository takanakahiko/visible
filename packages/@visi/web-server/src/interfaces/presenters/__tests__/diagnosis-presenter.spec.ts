import { diagnosis } from '../../../__fixtures__/diagnosis';
import { createContainer } from '../../../tests/container';
import { DiagnosisPresenter } from '../diagnosis-presenter';

describe('DiagnosisPresenter', () => {
  let diagnosisPresenter: DiagnosisPresenter;

  beforeAll(() => {
    const container = createContainer();
    diagnosisPresenter = container.get(DiagnosisPresenter);
  });

  it('transforms diagnosis into API response', () => {
    const response = diagnosisPresenter.run(diagnosis);
    expect(response).toMatchSnapshot();
  });
});
