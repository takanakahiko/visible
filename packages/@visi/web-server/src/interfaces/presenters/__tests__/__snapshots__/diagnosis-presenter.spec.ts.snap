// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagnosisPresenter transforms diagnosis into API response 1`] = `
Object {
  "createdAt": 2020-05-17T08:28:59.186Z,
  "doneCount": 1,
  "id": "08eecb12-75a1-4798-aca2-f9e919b1fd56",
  "screenshot": "https://example.com",
  "sources": Array [
    Object {
      "content": "<html></html>",
      "id": "08eecb12-75a1-4798-aca2-f9e919b1fd56",
      "reports": Array [
        Object {
          "diffHunk": undefined,
          "difficulty": undefined,
          "id": "08eecb12-75a1-4798-aca2-f9e919b1fd56",
          "impact": undefined,
          "location": undefined,
          "message": "Error!",
          "outcome": "FAIL",
          "screenshot": undefined,
          "target": "/html/body",
        },
      ],
      "url": "https://example.com",
    },
  ],
  "status": "STARTED",
  "totalCount": 1,
  "updatedAt": 2020-05-17T08:28:59.186Z,
  "url": "https://exmaple.com",
}
`;
